src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ValueProviderResult.cs

```diff
@@ -3,227 +3,372 @@
-                    // array + convert
-                    var element = ConvertSimpleType(culture, value, destinationElementType);
-                    var converted = (IList)Array.CreateInstance(destinationElementType, 1);
-                    converted[0] = element;
-                    return converted;
+                    return null;
                 }
             }
-            else if (valueAsArray != null)
+        }
+
+        /// <summary>
+        /// Gets the number of submitted values.
+        /// </summary>
+        public int Length
+        {
+            get
             {
-                // case 3: destination type is single element but source is array, so extract first element + convert
-                if (valueAsArray.Length > 0)
+                if (Values != null)
+                {
+                    return Values.Length;
+                }
+                else if (Value != null)
                 {
-                    value = valueAsArray.GetValue(0);
-                    return ConvertSimpleType(culture, value, destinationType);
+                    return 1;
                 }
                 else
                 {
-                    // case 3(a): source is empty array, so can't perform conversion
-                    return null;
+                    return 0;
                 }
             }
+        }
 
-            // case 4: both destination + source type are single elements, so convert
-            return ConvertSimpleType(culture, value, destinationType);
+        /// <inheritdoc />
+        public override bool Equals(object obj)
+        {
+            var other = obj as ValueProviderResult?;
+            return other.HasValue ? Equals(other.Value) : false;
         }
 
-        private object ConvertSimpleType(CultureInfo culture, object value, Type destinationType)
+        /// <inheritdoc />
+        public bool Equals(ValueProviderResult other)
         {
-            if (value == null || value.GetType().IsAssignableFrom(destinationType))
+            if (Length != other.Length)
             {
-                return value;
+                return false;
             }
+            else
+            {
+                var x = (string[])this;
+                var y = (string[])other;
+                for (var i = 0; i < x.Length; i++)
+                {
+                    if (!string.Equals(x[i], y[i], StringComparison.Ordinal))
+                    {
+                        return false;
+                    }
+                }
+                return true;
+            }
+        }
+
+        /// <inheritdoc />
+        public override int GetHashCode()
+        {
+            return ((string)this)?.GetHashCode() ?? 0;
+        }
+
+        /// <inheritdoc />
+        public override string ToString()
+        {
+            return (string)this;
+        }
+
+        /// <inheritdoc />
+        public Enumerator GetEnumerator()
```

Comment:
Where would this inherit docs from? Already have explicit implementations of both methods in `IEnumerable<string>`. Might want to make this method `private Enumerator GetMyEnumerator()`.